nicegui/outlet.py

```diff
@@ -0,0 +1,180 @@
+from typing import Callable, Any, Self, Optional, Generator, Union
+
+from nicegui.client import Client
+from nicegui.single_page_router import SinglePageRouter
+from nicegui.single_page_router_config import SinglePageRouterConfig
+from nicegui.elements.router_frame import RouterFrame
+from nicegui.single_page_target import SinglePageTarget
+
+PAGE_TEMPLATE_METHOD_NAME = "page_template"
+
+
+class Outlet(SinglePageRouterConfig):
+    """An outlet allows the creation of single page applications which do not reload the page when navigating between
+    different views. The outlet is a container for multiple views and can contain nested outlets.
+
+    To define a new outlet, use the @ui.outlet decorator on a function which defines the layout of the outlet.
+    The layout function must be a generator function and contain a yield statement to separate the layout from the
+    content area. The yield can also be used to pass properties to the content are by return a dictionary with the
+    properties. Each property can be received as function argument in all nested views and outlets.
+
+    Once the outlet is defined, multiple views can be added to the outlet using the @outlet.view decorator on
```

Comment:
I updated it accordingly. One very important point hit me though quite "right into my face" when we started building our first real app based upon this feature and this is actually independent of if one uses the OOP version or the function based one:

De facto the ui.outlets have the same goal as APIRouters had in FastAPI world or Blueprints in the Flask universe, so like a branch of pages which can have several views ("leafs") or further outlets ("branches") if you want so.

**BUT** the function we are actually decorate ... and which can in many cases be super irrelevant or just a single yield ... or... more importantly... be identical for multiple pages.... thus should be sharable.... between two completely different apps at different locations in the hierarchy.

Or with other words: After now spending "months" on this pull request and seing now how its actually being used in practice I think that a more "FastAPI"ish / non-decorater style might be more intuitive.

So instead of 
```
@ui.outlet("/mail")
def email_app():
    # ... header 
    yield # content
    # ... footer

@email_app.view()
def email_inbox():
    ....
    
@ui.outlet("/calendar")
def calendar_app():
    # ... header 
    yield # content
    # ... footer

@calendar_app.view()
def calendar_overview():
    ....
```

rather

```
def app_base_layout():
    # ... header 
    yield # content
    # ... footer

e_mail = ui.outlet(layout=app_base_layout)
@email_app.view()
def email_inbox():
    ....

calendar_app = ui.outlet(layout=app_base_layout)    
@calendar_app.view()
def calendar_overview():
    ....
```

or in object oriented style:

```
.... work in progress, lunch :)
```

Reply:
I think the outlet (with it's layout) are a bit different from FastAPIs APIRouters or Flasks blueprints. Especially if we get rid of the current destinction between outlet and views (see https://github.com/Alyxion/nicegui/pull/3).

To think about OOP is still worthwhile. Maybe with some inspiration from the controllers in Litestar: https://docs.litestar.dev/latest/#expanded-example...